packages/beacon-node/src/network/reqresp/handlers/blobsSidecarsByRange.ts

```diff
@@ -0,0 +1,91 @@
+import {IBeaconChain} from "../../../chain/index.js";
+import {IBeaconDb} from "../../../db/index.js";
+
+// TODO EIP-4844: Unit test
+
+export async function* onBlobsSidecarsByRange(
+  requestBody: eip4844.BlobsSidecarsByRangeRequest,
+  chain: IBeaconChain,
+  db: IBeaconDb
+): AsyncIterable<EncodedPayload<eip4844.BlobsSidecar>> {
+  const {startSlot} = requestBody;
+  let {count} = requestBody;
+
+  if (count < 1) {
+    throw new ResponseError(RespStatus.INVALID_REQUEST, "count < 1");
+  }
+  // TODO: validate against MIN_EPOCHS_FOR_BLOCK_REQUESTS
+  if (startSlot < GENESIS_SLOT) {
+    throw new ResponseError(RespStatus.INVALID_REQUEST, "startSlot < genesis");
+  }
+
+  if (count > MAX_REQUEST_BLOCKS) {
+    count = MAX_REQUEST_BLOCKS;
+  }
+
+  const endSlot = startSlot + count;
+
+  // SPEC: Clients MUST respond with blobs sidecars from their view of the current fork choice -- that is, blobs
+  // sidecars as included by blocks from the single chain defined by the current head. Of note, blocks from slots
+  // before the finalization MUST lead to the finalized block reported in the Status handshake.
+  // https://github.com/ethereum/consensus-specs/blob/11a037fd9227e29ee809c9397b09f8cc3383a8c0/specs/eip4844/p2p-interface.md#blobssidecarsbyrange-v1
+
+  const finalizedSlot = chain.forkChoice.getFinalizedBlock().slot;
+
+  // Finalized tram of blobs
+  // TODO EIP-4844: Should the finalized block be included here or below?
+
+  if (startSlot <= finalizedSlot) {
+    // Chain of blobs won't change
+    for await (const {key, value} of db.blobsSidecarArchive.binaryEntriesStream({gte: startSlot, lt: endSlot})) {
+      yield {
+        type: EncodedPayloadType.bytes,
+        bytes: value,
+        contextBytes: {
+          type: ContextBytesType.ForkDigest,
+          forkSlot: db.blockArchive.decodeKey(key),
+        },
+      };
+    }
+  }
+
+  // Non-finalized tram of blobs
+
+  if (endSlot > finalizedSlot) {
+    const headRoot = chain.forkChoice.getHeadRoot();
+    // TODO EIP-4844: forkChoice should mantain an array of canonical blocks, and change only on reorg
+    const headChain = chain.forkChoice.getAllAncestorBlocks(headRoot);
+
+    // Iterate head chain with ascending block numbers
+    for (const block of headChain) {
+      // Must include only blocks in the range requested
+      if (block.slot >= startSlot && block.slot < endSlot) {
+        // Note: Here the forkChoice head may change due to a re-org, so the headChain reflects the cannonical chain
+        // at the time of the start of the request. Spec is clear the chain of blobs must be consistent, but on
+        // re-org there's no need to abort the request
+        // Spec: https://github.com/ethereum/consensus-specs/blob/a1e46d1ae47dd9d097725801575b46907c12a1f8/specs/eip4844/p2p-interface.md#blobssidecarsbyrange-v1
+
+        const blockBytes = await db.blobsSidecar.getBinary(fromHex(block.blockRoot));
+        if (blockBytes) {
```

Comment:
if we don't have `blockBytes`, maybe throw Error otherwise the response of this is not in synchronized to `beaconBlocksByRange`

Reply:
Hmm, I'm undecided here, our db should never become out of sync with the fork-choice. However they could be some edge cases. Should we fail the entire response if we can't find the block in the db?

Reply:
I think we should, otherwise our response is not really a chain (one block may miss its parent for example) and the client node can't process all of the returned blocks

also found this in the spec https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#beaconblocksbyrange

```
Clients MUST respond with blocks that are consistent from a single chain within the context of the request.
```

Reply:
@tuyennhv then we should change the blocks by range implementation to also throw if the block binary is not found right?

Reply:
@dapplion  yeah I think so, gonna make a PR for it 👍